src/main/java/org/bytedeco/javacpp/tools/Parser.java

```diff
@@ -98,17 +98,40 @@ public Parser(Logger logger, Properties properties, String encoding, String line
     String lineSeparator = null;
     /** Java names of classes needing upcast from their subclasses. */
     Set<String> upcasts = new HashSet<>();
-    /** Classes that have a base class appearing as key in this map need a downcast constructor. The associated value gives the class to downcast from. */
-    Map<String, Set<Type>> downcasts = new HashMap<>();
+
+    static private class Inheritance {
+        Type base;
+        boolean virtual;
```

Comment:
Why can't we put this as a new `virtualInheritance` field in `Type`??

Reply:
Why do we need to clone anything? I'm not talking about reusing the `virtual`, I'm talking about adding a new one

Reply:
Because the same `Type` instance can be stored in `downcasts` several times with different value for virtualInheritance.

Reply:
Like maybe what we need isn't a new flag but a `Type virtualBase` field?

Reply:
Why would the same type have different properties? That doesn't make sense to me

Reply:
That's the parent class, the one we must downcast from, that is stored.

Reply:
So wouldn't we want a `List<Type> virtualSubclasses` or something as a field in `Type` instead?

Reply:
Anyway, I'm pretty sure we can model this more cleanly without adding a new `Inheritance` concept that doesn't map to anything concrete, but if you're OK to do it by cloning stuff around, please do that for now until this gets fixed properly later, somehow

Reply:
Let's take the example of opencv:

`cv::Algortithm` is polymorphic and we know some classes virtually inherit from it. So we add it an `upcast` info.
When we parse `Algorithm`, we spot the `upcast info` and we store in `downcasts` the association
`Algorithm` => [ `Algorithm`, false ] to remember that sublasses of `Algortihm` need a dowcast constructor from `Algorithm`, and that we no virtual inheritance is involved yet.

Then we parse `FaceRecognizer` that inherits non-virtually from Algorithm. We see the mapping in `downcasts`, so we add a new entry: `FaceRecognizer` => [`Algorithm`, false] so remember that subclasses of `FaceRecognizer` needs a downcast constructor from `Algorithm`, without virtual inheritance involved. 

Then we parse `Feature2D` that inherits virtually from Algorithm. We see the mapping in `downcasts`, so we add a new entry: `Feature2D` => [`Algorithm`, true] .

That's the same Type object stored 3 times.

The class `Inheritance` models the type of inheritance towards some parent class. That's quite concrete.

One alternative way to do this is without introducing this Inheritance class is probably to replace:
```
Map<String, Set<Inheritance>>
```
by:
```
Map<String, Map<Type, Boolean>>
```
Is that ok for you ?

Reply:
> The class `Inheritance` models the type of inheritance towards some parent class. That's quite concrete.

Right, so that's why I'm saying that it should be possible to model that more cleanly with Type only, have all the bits there

> Is that ok for you ?

That sounds OK as a temporary solution, yes, so we can do that for now